torch/lib/THCUNN/generic/VolumetricAdaptiveAveragePooling.cu

```diff
@@ -0,0 +1,173 @@
+
+
+  real *output_data;
+  real *input_data;
+
+  int64_t sizeD, isizeT, isizeH, isizeW;
+  int64_t istrideD, istrideT, istrideH, istrideW;
+  int64_t totalZ;
+
+  if (input->nDimension == 4) {
+    sizeD = input->size[0];
+    isizeT = input->size[1];
+    isizeH = input->size[2];
+    isizeW = input->size[3];
+
+    istrideD = input->stride[0];
+    istrideT = input->stride[1];
+    istrideH = input->stride[2];
+    istrideW = input->stride[3];
+
+    THCTensor_(resize4d)(state, output, sizeD, osizeT, osizeH, osizeW);
+
+    totalZ = sizeD * osizeT;
+  } else {
+    input = THCTensor_(newContiguous)(state, input);
+
+    int64_t sizeB = input->size[0];
+    sizeD = input->size[1];
+    isizeT = input->size[2];
+    isizeH = input->size[3];
+    isizeW = input->size[4];
+
+    istrideD = input->stride[1];
+    istrideT = input->stride[2];
+    istrideH = input->stride[3];
+    istrideW = input->stride[4];
+
+    THCTensor_(resize5d)(state, output, sizeB, sizeD, osizeT, osizeH, osizeW);
+
+    totalZ = sizeB * sizeD * osizeT;
+  }
+
+  input_data = THCTensor_(data)(state, input);
+  output_data = THCTensor_(data)(state, output);
+
+  int64_t offsetZ = 0;
+  dim3 threads(32, 8);
+  // each H*W plane is processed by blocksH thread blocks
+  int blocksH = max((int)(16L / totalZ), 1);
+  while (totalZ > 0) {
+    dim3 blocks(totalZ > 65535 ? 65535 : totalZ, blocksH);
+    cunn_VolumetricAdaptiveAveragePooling_updateOutput_kernel
+      <<<blocks, threads, 0, THCState_getCurrentStream(state)>>>(
+        input_data, output_data, isizeT, isizeH, isizeW, osizeT, osizeH, osizeW,
+        istrideD, istrideT, istrideH, istrideW, offsetZ
+      );
+
+    totalZ -= 65535;
+    offsetZ += 65535;
+    THCudaCheck(cudaGetLastError());
+  }
+
+  if (input->nDimension == 5) {
+    // clean
+    THCTensor_(free)(state, input);
+  }
+}
+
+void THNN_(VolumetricAdaptiveAveragePooling_updateGradInput)(
+           THCState *state,
+           THCTensor *input,
+           THCTensor *gradOutput,
+           THCTensor *gradInput)
+{
+  THCUNN_assertSameGPU(state, 3, input, gradOutput, gradInput);
+
+  gradOutput = THCTensor_(newContiguous)(state, gradOutput);
+
+  THCTensor_(resizeAs)(state, gradInput, input);
+  THCTensor_(zero)(state, gradInput);
+
+  real *gradInput_data;
+  real *gradOutput_data;
+
+  int64_t sizeD, isizeT, isizeH, isizeW;
+  int64_t osizeT, osizeH, osizeW;
+  int64_t totalZ;
+
+  if (input->nDimension == 4) {
+    sizeD = input->size[0];
+    isizeT = input->size[1];
+    isizeH = input->size[2];
+    isizeW = input->size[3];
+
+    osizeT = gradOutput->size[1];
+    osizeH = gradOutput->size[2];
+    osizeW = gradOutput->size[3];
+  } else {
+    sizeD = input->size[1];
+    isizeT = input->size[2];
+    isizeH = input->size[3];
+    isizeW = input->size[4];
+
+    osizeT = gradOutput->size[2];
+    osizeH = gradOutput->size[3];
+    osizeW = gradOutput->size[4];
+  }
+
+  // somehow nonatomic is passing all test for volumetric case.
```

Comment:
it should, assuming it is written correctly, right?  My understanding is that nonatomic means it's not using any atomic operations (e.g. atomicAdd); since it's operating over the input pixels this is okay because only one thread is writing to that location in the grad input.  The atomic version is operating over the output pixels; since an input can correspond to multiple outputs it needs to use atomic ops.

Reply:
Yeah, it should. Spatial version said its nonatomic kernel isn't passing the tests and thus disabled it. Either there is something weird going on with spatial kernel, or the author just copied from spatial adaptive max pooling, which also has that line (https://github.com/pytorch/pytorch/blob/master/torch/lib/THCUNN/generic/SpatialAdaptiveMaxPooling.cu#L97). But my volumetric adaptive max pooling also works for the nonatomic case in tests #2782 , so I'm not really sure what's going on.

Reply:
For the record, if I remember properly the `SpatialAdaptiveMaxPooling` didn't indeed pass the tests without atomic operations , so I just used it by default and didn't look back to see why it was the case :-)

Reply:
Thanks for clearing things up :)